### PR TITLE
sync our database with braintree

### LINKED ISSUE #3287: migrate to Braintree
https://www.braintreepayments.com/features/marketplace

<bountysource-plugin>
---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/10036780-migrate-to-braintree?utm_campaign=plugin&utm_content=tracker%2F85909&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F85909&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by whit537**
Closes #3287. Braintree has received data from Balanced. Now we need to knit the customer records in Braintree with our participants table.

### CODE REVIEW

File: sql/branch.sql [hunk 1]
```diff
@@ -0,0 +1,5 @@
+BEGIN;
+
+    ALTER TABLE participant ADD COLUMN braintree_id text NOT NULL DEFAULT '';
```
Comment (rohitpaulk) [inline sql/branch.sql]: We have a `braintree_customer_id` field already, is that what you were looking for?

Comment (whit537) [inline sql/branch.sql]: Ah, cool. :-)

**Comment by whit537** [comment]
I'm seeing a handful of cases where an absorbed account has a `balanced_customer_href`, and the account that absorbed it has the same one.

**Comment by whit537** [comment]
I also see [the](https://github.com/gratipay/gratipay.com/issues/3391#issuecomment-104411828) seven customers that Braintree knows about that we don't.

File: bin/branch.py [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+    assert row[CUSTOMER_ID] == row[CHECK], row
+
+    customer_id = row[CUSTOMER_ID]
+    card_id = row[CARD_ID]
+
+    p = db.one( "SELECT p.*::participants FROM participants p WHERE balanced_customer_href=%s"
+              , ("/customers/{}".format(customer_id),)
+               )
+
+    er = ExchangeRoute.insert(p, 'braintree-cc', card_id)
```
Comment (rohitpaulk) [inline bin/branch.py]: Before we do this, we should check whether a route exists for `braintree-cc`. If so, the user had updated his card, and we should just leave the participant as is.

File: bin/branch.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    db.run("UPDATE participants SET braintree_customer_id=%s", (customer_id,))
```
Comment (rohitpaulk) [inline bin/branch.py]: `WHERE` :smiley:

**Comment by whit537** [comment]
We want a clean slate with Braintree going forward. We want exactly one Braintree customer per open account. I'm not sure yet about how cards should sync up.

File: bin/branch.py (hunk 3 above)
Comment (whit537) [inline bin/branch.py]: ;-)

Still commented out tbh.

File: bin/branch.py (hunk 2 above)
Comment (whit537) [inline bin/branch.py]: Okay.

**Comment by rohitpaulk** [comment]
> We want exactly one Braintree customer per open account.

That's something I can handle next week as cleanup. As of now, we should be focusing on linking whatever `braintree_customer_id`s we have for people with valid cards.

**Comment by rohitpaulk** [comment]
One case where I see us having two braintree customers for one open account is https://github.com/gratipay/gratipay.com/pull/3473#discussion_r31285429. It's safe to delete the 'migrated' customer.

**Comment by whit537** [comment]
I'm seeing a number of closed and a few suspicious accounts with a `balanced_customer_href` that matches `customer_id` in the Braintree import log.

**Comment by whit537** [comment]
We've missed the bank payout deadline for today. We have so few users and everything is so disrupted, we can probably get away with slipping a few payouts until Monday. :-(

Let's land Braintree!

**Comment by whit537** [comment]
I also want to see which cards are `status: removed` at Balanced. :-(

**Comment by whit537** [comment]
It would be Bad(tm) if we started charging users on cards that they had disconnected. Kind of a wtf to even have those in the export, or at least not being invalidated at Braintree, if indeed that is what is happening.

**Comment by whit537** [comment]
Results using 8c34500fadb437fd89aae22b265579f935885c31 with the [import log](https://github.com/gratipay/logs/blob/master/misc/braintree-import/gratipay_import_log.csv) and the production database:

| status | participants |
|----------|-----------------:|
| ready | 4,201 |
| already migrated | 9 |
| closed | 247 |
| suspicious | 55 |
| archived | 48 |
| failed | 0 |
| total | 4,567 |

**Comment by whit537** [comment]
So now for each card_id in the `ready` set, we want to find out the status of that card on Balanced.

**Comment by whit537** [comment]
I'm not seeing `status` in the Balanced API, either the [docs](https://docs.balancedpayments.com/1.1/api/cards/#fetch-a-card) or live.

**Comment by whit537** [comment]
The closest I'm seeing is `can_{credit,debit}`.

**Comment by rohitpaulk** [comment]
Aren't cards deleted entirely when we call `unstore` on them?

**Comment by whit537** [comment]
```js
	status: function() {
		return this.get('can_debit') ? 'active' : 'removed';
	}.property('can_debit')
```

[src](https://github.com/balanced/balanced-dashboard/blob/900750621e480692f6e4d873a96214d5bf6da766/app/models/card.js#L50-L52)

**Comment by whit537** [comment]
> Aren't cards deleted entirely when we call `unstore` on them?

Is `unstore` a Balanced or Braintree API? I don't remember it from Balanced.

**Comment by whit537** [comment]
> Spot-checking, I'm seeing a card that is "status: removed" at Balanced that is merely "Expired" at Braintree. :open_mouth: 
>
https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/cards/CC9ZBPfVyWiXwRNKrwSaO4m
>
https://www.braintreegateway.com/merchants/msjyyvtsvvnc4ny6/payment_methods/any/CC9ZBPfVyWiXwRNKrwSaO4m

**Comment by rohitpaulk** [comment]
> Is unstore a Balanced or Braintree API? I don't remember it from Balanced.

https://github.com/gratipay/gratipay.com/blob/master/gratipay/models/exchange_route.py#L81

**Comment by whit537** [comment]
Okay. What do the docs say for it? I'm finding a bunch of `removed` cards, results forthcoming ...

**Comment by whit537** [comment]
Okay, this is effed up. Check out these (anonymized) screenshots from Balanced and Braintree. The *removed* card at Balanced is the *default* card at Braintree. The warning indicator at Braintree is simply because it expires next month. All indicators are that I could charge that card right now if I wanted to. O.O

screen shot 2015-05-28 at 8 47 57 pm

screen shot 2015-05-28 at 8 48 19 pm

**Comment by whit537** [comment]
Script still churning, 403 removed cards so far ...

**Comment by rohitpaulk** [comment]
:confused:

Should we run this through Braintree support, or handle it manually?

**Comment by whit537** [comment]
@rohitpaulk We have to handle it manually if we're going to run payday tonight. :(

**Comment by rohitpaulk** [comment]
Hmm, true. Our paydays are getting more eventful by the week :boom:

**Comment by whit537** [comment]
Yeah, and we're not through yet. :hatching_chick:

**Comment by whit537** [comment]
Okay! Where are we? I'm back with food, probably good for another ~3 hours at least.

**Comment by whit537** [comment]
@rohitpaulk I expect you'll have to tap out soon, get some :zzz:?

**Comment by rohitpaulk** [comment]
I'm good for around 2 hours. :)

**Comment by whit537** [comment]
Results of 631d8cdee6c51a1640705589e7e2d2fd53a1fb6e with output of https://github.com/gratipay/gratipay.com/pull/3473#issuecomment-106631148: 657 removed cards.

**Comment by whit537** [comment]
I think the hypothesis at https://github.com/gratipay/gratipay.com/issues/3391#issuecomment-104424101 may be true, that Balanced included invalidated cards in the export to Braintree but not to us.

**Comment by whit537** [comment]
For example, for `whit537`, I see 27 cards at Balanced (all `removed`), and 26 at Braintree. A seemingly arbitrary card is `default` Braintree.

**Comment by whit537** [comment]
For example, for `whit537`, I see 27 cards at Balanced (all `removed`), and 27 at Braintree. A seemingly arbitrary card is `default` Braintree:

https://dashboard.balancedpayments.com/#/marketplaces/MP12Xw5lL6iaILtqImIoroDL/customers/AC4s77PqpGjIP9C9C6SJlJ6u

https://www.braintreegateway.com/merchants/msjyyvtsvvnc4ny6/customers/AC4s77PqpGjIP9C9C6SJlJ6u

**Comment by whit537** [comment]
I find exactly one row for `whit537` in the https://github.com/gratipay/gratipay.com/pull/3473#issuecomment-106631148 output. It's a removed card, also included in https://github.com/gratipay/gratipay.com/pull/3473#issuecomment-106647951.

Now, the import log *does* include two different kinds of records: customers and cards. The customer rows have a card field, and that's what https://github.com/gratipay/gratipay.com/pull/3473#issuecomment-106631148 looks at; it ignores the dedicated card rows. Presumably the card rows contain all of the other (invalidated) cards?

**Comment by whit537** [comment]
Yeah, the "export for stripe" I have from Balanced includes two rows for https://www.braintreegateway.com/merchants/msjyyvtsvvnc4ny6/customers/AC5cegJ2OFccgexg1I4iwYYU, but the `funding_instrument_guid` is only present for the non-removed card; it's empty in the second row.

**Comment by whit537** [comment]
I just deleted one of my 27 cards at Braintree. Now I have 26. Simple as that, it would appear.

https://developers.braintreepayments.com/javascript+python/reference/request/payment-method/delete

**Comment by whit537** [comment]
Do we want to call this API for all remove cards?

https://developers.braintreepayments.com/javascript+python/reference/request/payment-method/delete

**Comment by rohitpaulk** [comment]
> Do we want to call this API for all remove cards?

Yep, makes sense.

**Comment by whit537** [comment]
> Yeah, the "export for stripe" I have from Balanced includes two rows for [...]

However, it includes multiple cards for me, which of course are all `removed`.

**Comment by whit537** [comment]
I lied. It has three entries for me, a credit card, a bank account, and a customer:

```
,AC4s77PqpGjIP9C9C6SJlJ6u,,,CC5Gy43LFmCBXBQ7IS3I2Tm2,
,AC4s77PqpGjIP9C9C6SJlJ6u,,,BA1RZX9aubiSyqnCKjlniA9I,
,AC4s77PqpGjIP9C9C6SJlJ6u,,,,
```

Interestingly, I remove the credit after taking the "export for stripe" a couple weeks ago. I've started a new export. I hypothesize that the `CC5Gy43LFmCBXBQ7IS3I2Tm2` card will no longer be present. Instead, I will have only two entries, one for the bank account and one for the customer.

**Comment by whit537** [comment]
Yesssssssssss.

```
,AC4s77PqpGjIP9C9C6SJlJ6u,,,BA1RZX9aubiSyqnCKjlniA9I,
,AC4s77PqpGjIP9C9C6SJlJ6u,,,,
```

**Comment by whit537** [comment]
Alright! So this `export-for-stripe` gives us the correct mapping of `customer_id` to `card_id`.

I reviewed the [Technical Migration Guide](https://www.balancedpayments.com/stripe/migration-guide) and didn't notice anything super-relevant.

**Comment by rohitpaulk** [comment]
> Alright! So this export-for-stripe gives us the correct mapping of customer_id to card_id.

We still need to remove all the invalidated (balanced) cards from Braintree. And once we've done that, we won't need the 'correct mapping', coz that'll correspond to the only card(s) left, correct?

**Comment by whit537** [comment]
@rohitpaulk Well, I'm thinking that we should iterate over all cards at Braintree and delete the ones that aren't in the export-for-stripe, because https://github.com/gratipay/gratipay.com/pull/3473#issuecomment-106647951 only looked at cards that were mentioned in a customer row in the import log, ignoring cards mentioned in a card row.

**Comment by whit537** [comment]
The new `export-for-stripe` I just took removes a handful of payment methods, without adding any. Is that consistent with our vaulting changes, @rohitpaulk? We're not storing any new cards at Balanced anymore, are we?

**Comment by whit537** [comment]
We don't want to clobber cards that have been newly added to Braintree.

**Comment by rohitpaulk** [comment]
> We're not storing any new cards at Balanced anymore, are we?

No, we're not. 

> We don't want to clobber cards that have been newly added to Braintree.

Yes, the check in the script should ensure that.

**Comment by rohitpaulk** [comment]
> Well, I'm thinking that we should iterate over all cards at Braintree and delete the ones that aren't in the export-for-stripe

All cards? Or only the ones in the import log? If it's all cards, we're going to lose the new cards added.

**Comment by rohitpaulk** [comment]
I somehow feel that we should use the real sources of truth, and not the `export-for-stripe`. 

- Iterate over all the removed cards on Balanced, and if they exist on Braintree, delete them. 
- Then iterate over the braintree customers from the import log, but instead of using the `card_id` from there, pick it off of braintree's API. (`customer.cards[0]`)

@whit537 - How sure are we about `export-for-stripe` only containing valid cards, and that it contains **all** valid cards?

**Comment by whit537** [comment]
API calls are way more expensive than reading local CSVs. :-/

**Comment by rohitpaulk** [comment]
`threaded_map` :smiley:

**Comment by whit537** [comment]
As long as we don't run on DO. :P

**Comment by whit537** [comment]
Some pseudocode:

```python
for card in all_cards_in_braintree:
    if card.customer is unknown:
        if there_is_another_way_to connect_back_to_a_participant:
            connect_back_to_a_participant()
        else:
            continue        
    if card.removed_at_balanced:
        card.delete()
    if card.customer.has_other_card_at_braintree:
        continue
```

**Comment by whit537** [comment]
I think we should iterate over a fresh Customers & Payment Methods export from Braintree.

screen shot 2015-05-28 at 10 36 07 pm

**Comment by rohitpaulk** [comment]
:+1:

**Comment by whit537** [comment]
0cc4896ecb8eeedee117e70f0f93aa9131b077ad ran in ~11 minutes on my laptop.

**Comment by whit537** [comment]
Results of https://github.com/gratipay/gratipay.com/commit/9b39bffee047abf75cbe2a399fdd9c3c0e740b68 on a Braintree export from two weeks ago:

| status       | cards |
|:---------------|--------:|
| ready to sync |  3,619 |
| removed at Balanced |  1,649 |
| orphaned |    327 |
| closed | 172 |
| suspicious | 46 |
| archived | 41 |
| migrated themselves | 9 |
| unknown | 2 |
| total | 5,865 |

**Comment by whit537** [comment]
My next steps are to run the script with a fresh export from Braintree, and then uncomment the live parts, test it against on or two records, then let it loose.

**Comment by whit537** [comment]
With orphan recovery (fa9525ef9c9d10ce2c4efbd58fd8813fd7b221ad):

| status       | cards |
|:---------------|--------:|
| synced |  3,619 |
| deleted |  2,246 |
| total | 5,865 |

Without orphan recovery (9b39bffee047abf75cbe2a399fdd9c3c0e740b68):

| status       | cards |
|:---------------|--------:|
| synced |  3,619 |
| deleted |  1,919 |
| orphaned |  327 |
| total | 5,865 |

No orphans are recoverable.

**Comment by whit537** [comment]
Running b34b3b695354ef5be1fecdb0abfb018231ab1447 on a fresh export ...

**Comment by whit537** [comment]
Alright! Once this is done, it's time to run for real. O.O

**Comment by whit537** [comment]
:athletic_shoe:

**Comment by whit537** [comment]
Results of b34b3b6 on a fresh export:

| status       | cards |
|:---------------|--------:|
| synced |  3,619 |
| deleted |  2,245 |
| total | 5,864 |

Okay! Balanced hasn't exported any more cards, we have't created any with a `CC` prefix, and we've only deleted once card: https://github.com/gratipay/gratipay.com/pull/3473#issuecomment-106652192. <:-)

**Comment by whit537** [comment]
I fiddled with the script to exit after a single delete. It appears to have worked!

**Comment by whit537** [comment]
I just satisfied myself that the `get_participant_by_card_id` sql *does* work.

**Comment by rohitpaulk** [comment]
:rocket:

**Comment by whit537** [comment]
:eyes:

**Comment by whit537** [comment]
I'm testing a single sync. Then I'm ready to run it for real!

**Comment by whit537** [comment]
Okay! I'm seeing `participant_id` set properly at Braintree, and the `braintree_customer_id` and exchange route in Gratipay. Ready for lift off ...

**Comment by whit537** [comment]
Last chance for review ...

**Comment by whit537** [comment]
/me waits for Travis on principle ...

**Comment by whit537** [comment]
I'm expecting this script to run for 15 or 20 minutes, since there's more network traffic with the additional API and database calls.

**Comment by whit537** [comment]
Okay! :vertical_traffic_light:

**Comment by whit537** [comment]
Script started.

**Comment by whit537** [comment]
```
Traceback (most recent call last):
  File "/Users/whit537/personal/gratipay/gratipay.com/gratipay/billing/payday.py", line 44, in g
    return func(*a, **kw)
  File "bin/sync-with-braintree.py", line 120, in process_record
    braintree.PaymentMethod.delete(card_id)
  File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/braintree/payment_method
    return Configuration.gateway().payment_method.delete(payment_method_token)
  File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/braintree/payment_method
    self.config.http().delete("/payment_methods/any/" + payment_method_token)
  File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/braintree/util/http.py",
    return self.__http_do("DELETE", path)
  File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/braintree/util/http.py",
    Http.raise_exception_from_status(status)
  File "/Users/whit537/personal/gratipay/gratipay.com/env/lib/python2.7/site-packages/braintree/util/http.py",
    raise NotFoundError()
NotFoundError
```

**Comment by rohitpaulk** [comment]
Hmm, only delete if the card exists? We'll also have to patch up all the other inserts/deletes to not repeat themselves if this wasn't the first record or so.

**Comment by whit537** [comment]
@rohitpaulk There was only one sync but a few deletes. Fortunately deletes are easier to work around.

**Comment by whit537** [comment]
Script restarted ...

**Comment by whit537** [comment]
At the end of this script, `whit537` should have 0 cards attached at Braintree, though the customer should still exist.

**Comment by whit537** [comment]
Down from 26 to 12 so far. :-)

**Comment by whit537** [comment]
:haircut:

**Comment by whit537** [comment]
2 left!

**Comment by whit537** [comment]
Okaaaaaay, more than 20 minutes. :clock230:

**Comment by whit537** [comment]
No more :credit_card: for `whit537`.

**Comment by whit537** [comment]
```
real    25m32.430s
user    5m18.790s
sys     0m33.644s
```

**Comment by whit537** [comment]
So now a Customers & Payment Methods export from Braintree should have 3,619 records, right?

**Comment by whit537** [comment]
Sigh. 3,618. :-/

Oh well, close enough!

!m @rohitpaulk

**Comment by rohitpaulk** [comment]
!m @whit537

**Comment by whit537** [comment]
Log uploaded in https://github.com/gratipay/logs/commit/b9860556d64c5b8980494267456e8efc8478e451.

**Pull request closed without merging**